fix: Account for group deletes when indexing groupType

tomasfarias opened the pull request:
## Problem

When we delete a `GroupTypeMapping` object via the admin UI, the `group_type_index` of the remaining objects is not adjusted to account for the deletion of the `GroupTypeMapping`. This makes it so the actual index (now with one less item) doesn't match `group_type_index`.

In the frontend, we do not account for possible deletes, thus causing us to default to "Persons" as the label.

<!-- Who are we building for, what are their needs, why is this important? -->

## Changes

Check for the `group_type_index` value instead of assuming it matches the array.

I thought about doing the significantly more elegant `group_type_index % groupTypes.length()` but it doesn't quite work in all cases.

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Locally, start with initial set of groups:

![image](https://github.com/PostHog/posthog/assets/18740659/4fdea895-6751-4fad-93f7-17a32ba01437)

Delete Organization(s) (index 1) via the admin UI:

![image](https://github.com/PostHog/posthog/assets/18740659/031ad9d9-58b5-4366-a142-227058303016)

The local development instance rapidly created Organization(s) again, but now it has index 3 instead of 1.

Notice two things:
1. Organization(s) is now named "Persons" (default)
2. Instance(s) is now named "Organization(s)" (because Organization(s) now has index 2 in the array!)

If we apply my changes, things look now as expected:

![image](https://github.com/PostHog/posthog/assets/18740659/08d47eea-6a23-487d-a9ac-0c8b145b1eaa)

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

tomasfarias committed:
fix: Account for group deletes when indexing groupType

neilkakkar commented:
Makes sense!

Worth checking if anywhere else is making the same mistake. I browsed through the code, and while it _seems_ fine, I'd recommend making sure by:

1. Renaming `groupTypes` to `groupTypesArray`, see where ts complains and confirming things are used properly there.
2. Potentially change it from an array to dictionary to make sure the same problem isn't possible in the future. This will make sure deletes don't end up every changing the index of the group in question.

tomasfarias commented on frontend/src/layout/navigation/TopBar/TopBar.tsx:
```diff
@@ -39,7 +39,7 @@ export function TopBar(): JSX.Element {
     const hasNotebooks = !!featureFlags[FEATURE_FLAGS.NOTEBOOKS]
 
-    const groupTypes = [
+    const groupTypesArray = [
```
I'll revert this file 👍

tomasfarias replied:
Hmm, nevermind, doing that produces a Type error:

> Property 'groupTypes' does not exist on type 'IntrinsicAttributes & UniversalSearchPopoverProps<TaxonomicFilterValue>'. Did you mean 'groupType'?

neilkakkar commented on frontend/src/lib/components/UniversalSearch/UniversalSearchPopover.tsx:
```diff
@@ -32,7 +32,7 @@ export interface UniversalSearchPopoverProps<ValueType = TaxonomicFilterValue>
     value?: ValueType
     onChange?: (value: ValueType, groupType: TaxonomicFilterGroupType, item: SearchDefinitionTypes) => void
-    groupTypes?: TaxonomicFilterGroupType[]
+    groupTypesArray?: TaxonomicFilterGroupType[]
```
@tomasfarias I suspect the type error exists because we renamed it here too 😅 .

I think there are only 3-4 places that actually use groupTypes for Group Analytics Group types; rest are all TaxonomicFilter related, we can revert all of these

neilkakkar replied:
one by one I'm afraid, buuut that's what the rename helps with, if you just change `groupsModel`, and then run `pnpm typescript:check` it will show you all places that referenced it which you now need to change; and don't need to bother about the rest.

So revert all but the `groupsModel` file, see where ts errors pop up, rename those places, confirm they're not accessing by index, and we're done 👍

tomasfarias replied:
Ok, I've renamed all the files that popped up in `pnpm typescript:check`. Thank you!

tomasfarias replied:
Nevermind, now tests are failing...

Twixes reviewed (commented):
This fix only applies to `aggregationLabel`, right? Because `groupTypesArray` vel `groupTypes` is still loaded and set the same way.
(Note: the rename makes this PR quite a bit harder to read, and it doesn't really change anything, because in any case that reducer has an array value.)

What if instead `groupTypes` became `groupTypesRaw`, and we added a `groupTypes` selector that would have this value:
```typescript
const groupTypes: GroupType[] = new Array(groupTypesRaw.length)
for (const groupType of groupTypesRaw) {
    groupTypes[groupType.group_type_index] = groupType
}
return groupTypes
```
This way the `groupTypes` array would work as expected everywhere.

tomasfarias commented:
> This fix only applies to `aggregationLabel`, right?

Yeah, just wanted to fix the label.

> (Note: the rename makes this PR quite a bit harder to read, and it doesn't really change anything, because in any case that reducer has an array value.)

Yes.

> What if instead `groupTypes` became `groupTypesRaw`, and we added a `groupTypes` selector that would have this value:
> 
> ```ts
> const groupTypes: GroupType[] = new Array(groupTypesRaw.length)
> for (const groupType of groupTypesRaw) {
>     groupTypes[groupType.group_type_index] = groupType
> }
> return groupTypes
> ```
> 
> This way the `groupTypes` array would work as expected everywhere.

How would this work on calls to `groupTypes.map` after we have deleted some groups from the array, e.g.:

```
return groupTypesArray.map(
  groupType: GroupType) => `${TaxonomicFilterGroupType.GroupsPrefix}_${groupType.group_type_index}` as unknown as TaxonomicFilterGroupType
)
```

Won't map apply the function on an undefined value? Doesn't this require us to skip over undefined items in the array?

Personally I would go back to my first commit and revert everything else...

Twixes commented:
> Won't map apply the function on an undefined value?

No, uninitialized values are "empty" so they are skipped. You can test this with:
```typescript
let arr = new Array(5)
arr[0] = 'a'
arr[2] = 'c'
console.log(arr.map((val, i) => `${val} - ${i}`))
```

Twixes commented:
Though _ideally_ `groupTypes ` would be an object – that would just require us doing `Object.values(groupTypes).map(...)` instead of `groupTypes.map(...)`, and otherwise no other changes should be needed

tomasfarias commented:
> > Won't map apply the function on an undefined value?
> 
> No, uninitialized values are "empty" so they are skipped. You can test this with:

TIL: About js sparse arrays. Thank you!

tomasfarias committed:
refactor: Use a groupTypes selector to find existing group types

tomasfarias commented:
> What if instead groupTypes became groupTypesRaw, and we added a groupTypes selector that would have this value:

@Twixes I gave your solution a shot. Also had to add a `groupTypesLoading` selector as that's also imported in a lot of places.

neilkakkar commented:
yeah much nicer than renaming + manually checking usage 👍

Twixes reviewed (approved):
This is nice
